In [ ]:
import pandas as pd
df = pd.read_csv("HR_comma_sep.csv")
df.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

Data Exploration to understand the data features

In [ ]:
left_company = df[df.left == 1]
left_company.shape

(3571, 10)

In [ ]:
retained = df[df.left == 0]
retained.shape

(11428, 10)

In [ ]:
df.groupby('left').mean()

<ipython-input-197-b0959f325c8f>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('left').mean()


satisfaction_level  last_evaluation  number_project  \
left                                                        
0               0.666810         0.715473        3.786664   
1               0.440098         0.718113        3.855503   

      average_montly_hours  time_spend_company  Work_accident  \
left                                                            
0               199.060203            3.380032       0.175009   
1               207.419210            3.876505       0.047326   

      promotion_last_5years  
left                         
0                  0.026251  
1                  0.005321

Data Visulization


In [ ]:
# pd.crosstab(df.salary, df.left).bar(kind='bar')
pd.crosstab(df.salary, df.left)
# Candidatate who have high salary are likely to be retained in the company

left       0     1
salary            
high    1155    82
low     5144  2172
medium  5129  1317

In [ ]:
# pd.crosstab(df.Department, df.left).plot(kind='bar')
pd.crosstab(df.Department, df.left)
# There is not so high impact of department on employee leaving the company. So remove department section from analysis

left            0     1
Department             
IT            954   273
RandD         666   121
accounting    563   204
hr            524   215
management    539    91
marketing     655   203
product_mng   704   198
sales        3126  1014
support      1674   555
technical    2023   697

From the data analysis we conclude that we will use following variables as independant variables in our model.

1) **Satisfaction Level**

2) **Average Monthly Hours**

3) **Promotion Last 5 Years**

4) **Salary**


In [ ]:
Indep_df = df[['satisfaction_level','average_montly_hours','promotion_last_5years','salary']]
Indep_df.head()

satisfaction_level  average_montly_hours  promotion_last_5years  salary
0                0.38                   157                      0     low
1                0.80                   262                      0  medium
2                0.11                   272                      0  medium
3                0.72                   223                      0     low
4                0.37                   159                      0     low

In [ ]:
#Convert the salary variable to numbers for that we need to use dummy variable for that which can be got from pandas get_dummies function
salary_dummies = pd.get_dummies(Indep_df.salary, prefix ="salary")
salary_dummies

salary_high  salary_low  salary_medium
0                0           1              0
1                0           0              1
2                0           0              1
3                0           1              0
4                0           1              0
...            ...         ...            ...
14994            0           1              0
14995            0           1              0
14996            0           1              0
14997            0           1              0
14998            0           1              0

[14999 rows x 3 columns]

In [ ]:
#Concatinate the salary_dummies with the df we have to use. \
# Also we need to remove the text part of the salary variable as now we have the dummies for salary
df_withDummies = pd.concat([Indep_df, salary_dummies], axis = "columns")
df_withDummies

satisfaction_level  average_montly_hours  promotion_last_5years  \
0                    0.38                   157                      0   
1                    0.80                   262                      0   
2                    0.11                   272                      0   
3                    0.72                   223                      0   
4                    0.37                   159                      0   
...                   ...                   ...                    ...   
14994                0.40                   151                      0   
14995                0.37                   160                      0   
14996                0.37                   143                      0   
14997                0.11                   280                      0   
14998                0.37                   158                      0   

       salary  salary_high  salary_low  salary_medium  
0         low            0           1              0  
1      medium            0           0              1  
2      medium            0           0              1  
3         low            0           1              0  
4         low            0           1              0  
...       ...          ...         ...            ...  
14994     low            0           1              0  
14995     low            0           1              0  
14996     low            0           1              0  
14997     low            0           1              0  
14998     low            0           1              0  

[14999 rows x 7 columns]

In [ ]:
df_withDummies.drop(['salary'], axis='columns', inplace=True)
df_withDummies.head()


satisfaction_level  average_montly_hours  promotion_last_5years  \
0                0.38                   157                      0   
1                0.80                   262                      0   
2                0.11                   272                      0   
3                0.72                   223                      0   
4                0.37                   159                      0   

   salary_high  salary_low  salary_medium  
0            0           1              0  
1            0           0              1  
2            0           0              1  
3            0           1              0  
4            0           1              0

In [ ]:
X = df_withDummies
X

satisfaction_level  average_montly_hours  promotion_last_5years  \
0                    0.38                   157                      0   
1                    0.80                   262                      0   
2                    0.11                   272                      0   
3                    0.72                   223                      0   
4                    0.37                   159                      0   
...                   ...                   ...                    ...   
14994                0.40                   151                      0   
14995                0.37                   160                      0   
14996                0.37                   143                      0   
14997                0.11                   280                      0   
14998                0.37                   158                      0   

       salary_high  salary_low  salary_medium  
0                0           1              0  
1                0           0              1  
2                0           0              1  
3                0           1              0  
4                0           1              0  
...            ...         ...            ...  
14994            0           1              0  
14995            0           1              0  
14996            0           1              0  
14997            0           1              0  
14998            0           1              0  

[14999 rows x 6 columns]

In [ ]:
Y = df.left
Y

0        1
1        1
2        1
3        1
4        1
        ..
14994    1
14995    1
14996    1
14997    1
14998    1
Name: left, Length: 14999, dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, train_size=0.9)



In [ ]:
from sklearn.linear_model import LogisticRegression
analytics_model = LogisticRegression()


In [ ]:
analytics_model.fit(X_train, Y_train)

LogisticRegression()

In [ ]:
analytics_model.predict(X_test)

array([0, 0, 0, ..., 0, 0, 0])

Testing the Accuracy of our model

In [ ]:
analytics_model.score(X_test, Y_test)

0.788

We noticed that the accuracy is just 78.8% which is low. It means that there are other factors that are affecting the people leaving the company. Maybe by taking the Department wise consideration, we will have a better accuracy. That task is for later.